### PR TITLE
DEVX support

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8013/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8014/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe 
(still pending upstream support merge and FW support)

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+
+#ifndef UCT_IB_MLX5_IFC_H_
+#define UCT_IB_MLX5_IFC_H_
+
+#include <endian.h>
+#include <linux/types.h>
+
+#define u8 uint8_t
+
+#define __uct_nullp(typ) ((struct mlx5_ifc_##typ##_bits *)0)
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: coding style: macro parameters start with `_`

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+	u8         sho[0x1];
+	u8         tph[0x1];
+	u8         rf[0x1];
+	u8         dct[0x1];
+	u8         qos[0x1];
+	u8         eth_net_offloads[0x1];
+	u8         roce[0x1];
+	u8         atomic[0x1];
+	u8         reserved_at_21f[0x1];
+
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: what does this empty lines separate? Did not find a pattern

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+union mlx5_ifc_hca_cap_union_bits {
+	struct mlx5_ifc_cmd_hca_cap_bits cmd_hca_cap;
+	u8         reserved_at_0[0x8000];
+};
+
+struct mlx5_ifc_query_hca_cap_out_bits {
+	u8         status[0x8];
+	u8         reserved_at_8[0x18];
+
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: extra line here and in some places below

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+#define UCT_GET(typ, p, fld) ((be32toh(*((__be32 *)(p) +\
+                                        __uct_dw_off(typ, fld))) >> __uct_dw_bit_off(typ, fld)) & \
+                __uct_mask(typ, fld))
+
+#define __UCT_SET64(typ, p, fld, v) do { \
+        BUILD_BUG_ON(__uct_bit_sz(typ, fld) != 64); \
+        *((__be64 *)(p) + __uct_64_off(typ, fld)) = htobe64(v); \
+} while (0)
+
+#define UCT_SET64(typ, p, fld, v) do { \
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: what is the point to wrap `__UCT_SET64`?

File: src/uct/ib/mlx5/ib_mlx5_ifc.h (hunk 2 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/ib_mlx5_ifc.h]: dwords (32 bit chunks)

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 5]
```diff
@@ -7,78 +7,78 @@
...
-#define UCT_GET(typ, p, fld) ((be32toh(*((__be32 *)(p) +\
-                                        __uct_dw_off(typ, fld))) >> __uct_dw_bit_off(typ, fld)) & \
-                __uct_mask(typ, fld))
-
-#define __UCT_SET64(typ, p, fld, v) do { \
-        BUILD_BUG_ON(__uct_bit_sz(typ, fld) != 64); \
-        *((__be64 *)(p) + __uct_64_off(typ, fld)) = htobe64(v); \
-} while (0)
+#define UCT_GET(_typ, _p, _fld) \
+                ((be32toh(*((__be32 *)(_p) + \
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: indentation here and in other places (we use 4 symbols shifts)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8080/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8086/ for details (Mellanox internal link).

**Comment by brminich** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8095/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8124/ for details (Mellanox internal link).

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 6]
```diff
@@ -70,10 +71,55 @@ static ucs_status_t uct_ib_mlx5_device_init(uct_ib_device_t *dev)
...
+    ret = mlx5dv_devx_general_cmd(ctx, in, sizeof(in), out, sizeof(out));
+    if (ret == 0) {
+        if (UCT_GET(query_hca_cap_out, out, capability.cmd_hca_cap.dct)) {
+            dev->flags |= UCT_IB_DEVICE_FLAG_DC;
+        }
+        if (UCT_GET(query_hca_cap_out, out,
+                    capability.cmd_hca_cap.compact_address_vector)) {
+            dev->flags |= UCT_IB_DEVICE_FLAG_AV;
+        }
+    } else if (errno != EPERM &&
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: ( ) around conditions

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+#define UCT_ST_SZ_BYTES(_typ) (sizeof(struct mlx5_ifc_##_typ##_bits) / 8)
+#define UCT_ST_SZ_DW(_typ) (sizeof(struct mlx5_ifc_##_typ##_bits) / 32)
+#define UCT_ST_SZ_QW(_typ) (sizeof(struct mlx5_ifc_##_typ##_bits) / 64)
+#define UCT_UN_SZ_BYTES(_typ) (sizeof(union mlx5_ifc_##_typ##_bits) / 8)
+#define UCT_UN_SZ_DW(_typ) (sizeof(union mlx5_ifc_##_typ##_bits) / 32)
+#define UCT_BYTE_OFF(_typ, _fld) (__uct_bit_off(_typ, _fld) / 8)
+#define UCT_ADDR_OF(_typ, _p, _fld) ((char *)(_p) + UCT_BYTE_OFF(_typ, _fld))
+
+/* insert a value to a struct */
+#define UCT_SET(_typ, _p, _fld, _v) \
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: pls use UCT_IB_MLX5DV prefix for all macros

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+do { \
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: indentation (for all macros)

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+    UCS_STATIC_ASSERT(__uct_st_sz_bits(_typ) % 32 == 0); \
+    *((__be32 *)(_p) + __uct_dw_off(_typ, _fld)) = \
+    htobe32((be32toh(*((__be32 *)(_p) + __uct_dw_off(_typ, _fld))) & \
+        (~__uct_dw_mask(_typ, _fld))) | ((__uct_mask(_typ, _fld)) \
+        << __uct_dw_bit_off(_typ, _fld))); \
+} while (0)
+
+enum {
+	MLX5_CMD_OP_QUERY_HCA_CAP                 = 0x100
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: UCT_IB_MLX5

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+};
+
+enum mlx5_cap_mode {
+	HCA_CAP_OPMOD_GET_MAX	= 0,
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: UCT_IB_MLX5

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+	HCA_CAP_OPMOD_GET_CUR	= 1
+};
+
+enum mlx5_cap_type {
+	MLX5_CAP_GENERAL = 0
+};
+
+struct mlx5_ifc_cmd_hca_cap_bits {
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: uct_ib_mlx5

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 12]
```diff
@@ -70,10 +71,55 @@ static ucs_status_t uct_ib_mlx5_device_init(uct_ib_device_t *dev)
...
+
+    return status;
+}
+
 UCT_IB_DEVICE_INIT(uct_ib_mlx5_device_init);
 
 #endif
 
+#if HAVE_DECL_MLX5DV_DEVX_GENERAL_CMD
+ucs_status_t uct_ib_mlx5_get_compact_av(uct_ib_iface_t *iface, int *compact_av)
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: maybe remove this function altogether, and check UCT_IB_DEVICE_FLAG_AV during non-dv device init as well (the old way)?
also need to rename uct_ib_mlx5_device_init() to uct_ib_mlx5dv_device_init()

**Pull request closed without merging**